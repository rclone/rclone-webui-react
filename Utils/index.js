import checkPropTypes from 'check-prop-types';
import {applyMiddleware, createStore} from 'redux';
import rootReducer from './../src/reducers';
import {middleware} from '../src/store';

export const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
};

export const checkProps = (component, expectedProps) => {
    const propsErr = checkPropTypes(component.propTypes, expectedProps, 'prop', component.name);
    return propsErr;
};

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};