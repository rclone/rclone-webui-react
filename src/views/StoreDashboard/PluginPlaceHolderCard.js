import {Button, Card, CardBody, Col, Row} from "reactstrap";
import React, {useState} from "react";
import * as PropTypes from "prop-types";
import {toast} from "react-toastify";
import axiosInstance from "../../utils/API/API";


function PluginPlaceHolderCard({plugin}) {
    const {
        name, description, author, longDescription, icon, url, repo
    } = plugin;
    const [isLoading, setIsLoading] = useState(false);


    const activatePlugin = (e) => {
        setIsLoading(true);

        axiosInstance.post("pluginsctl/addPlugin", {
            url: repo
        }).then(res => {
            setIsLoading(false)
            toast.info(`Plugin ${name} added`);
        }, err => {
            setIsLoading(false)

            toast.error(`An error occurred: ${err}`)
        })
    }

    return (
        <Card data-test="pluginPlaceHolder">
            <CardBody>
                <Row>
                    <Col lg={3} className="d-none d-md-block">
                        <img style={{width: "100%"}} alt={icon ? name + " Icon" : 'rclone logo'}
                             src={icon ? icon : require("../../assets/img/brand/logo_symbol.png")}/>
                    </Col>
                    <Col lg={6} sm={12} md={6}>
                        <h3>{description}</h3>

                        <p style={{
                            textOverflow: "...",
                            overflow: "hidden",
                            maxHeight: "9.0em",
                            lineHeight: "1.8em"
                        }}>{longDescription}</p>

                        <p>By {author}</p>
                    </Col>
                    <Col lg={3} md={3}>
                        {isLoading ? "Installing..." :
                            <>
                                <Button color={"primary mb-2"} onClick={activatePlugin}>Activate</Button>
                                <Button color={"link p-sm-2 p-lg-0"}
                                        onClick={(e) => url ? window.open(url) : toast.error("Url not specified for details")}>More
                                    Details</Button>
                            </>
                        }

                    </Col>
                </Row>
            </CardBody>
        </Card>
    )
}

PluginPlaceHolderCard.propTypes = {
    plugin: PropTypes.shape({
        name: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        author: PropTypes.string.isRequired,
        longDescription: PropTypes.string.isRequired,
        icon: PropTypes.string,
        repo: PropTypes.string.isRequired,
        bugs: PropTypes.string,
    }),
    getPlugins: PropTypes.func.isRequired
}

export default PluginPlaceHolderCard