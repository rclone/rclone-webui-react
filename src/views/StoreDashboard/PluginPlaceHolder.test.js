import React from "react";
import {shallow} from "enzyme";
import toJson from "enzyme-to-json";
import {findByTestAttr, testStore} from "../../../Utils";
import PluginPlaceHolderCard from "./PluginPlaceHolderCard";

const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);

    const component = shallow(<PluginPlaceHolderCard {...props} store={store}/>);
    return component;
};


describe('Plugin place holder', function () {
    describe('renders', function () {
        let wrapper;
        beforeEach(() => {
            const initialState = {};

            const props = {
                plugin: {
                    "name": "Video Player",
                    "description": "A video player for rclone",
                    "author": "negative0",
                    "longDescription": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
                    "icon": "",
                    "repo": "https://github.com/negative0/rclone-video-plugin/",
                    "bugs": "https://github.com/negative0/rclone-video-plugin/issues"
                },
                getPlugins: jest.fn()

            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            const component = findByTestAttr(wrapper, "pluginPlaceHolder");
            expect(component).toHaveLength(1);
        });

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot();
        });

    });
});
