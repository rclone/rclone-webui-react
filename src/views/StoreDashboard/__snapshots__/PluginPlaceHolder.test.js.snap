// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin place holder renders should match snapshot 1`] = `
<Card
  data-test="pluginPlaceHolder"
  tag="div"
>
  <CardBody
    tag="div"
  >
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Col
        className="d-none d-md-block"
        lg={3}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <img
          alt="rclone logo"
          src="logo_symbol.png"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </Col>
      <Col
        lg={6}
        md={6}
        sm={12}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <h3>
          A video player for rclone
        </h3>
        <p
          style={
            Object {
              "lineHeight": "1.8em",
              "maxHeight": "9.0em",
              "overflow": "hidden",
              "textOverflow": "...",
            }
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
        <p>
          By 
          negative0
        </p>
      </Col>
      <Col
        lg={3}
        md={3}
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Button
          color="primary mb-2"
          onClick={[Function]}
          tag="button"
        >
          Activate
        </Button>
        <Button
          color="link p-sm-2 p-lg-0"
          onClick={[Function]}
          tag="button"
        >
          More Details
        </Button>
      </Col>
    </Row>
  </CardBody>
</Card>
`;
